---
title: "Variant attributes"
author: "Brian J. Knaus"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Subset to one chromosome}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



```{r}
library(vcfR)
data(vcfR_example)
```


## INFO


```{r}
mINFO <- metaINFO2df(vcf)
mINFO[1:3,1:4]
nrow(mINFO)
```



```{r}
INFO <- INFO2df(vcf)
head(INFO)
apply(INFO, MARGIN = 2, class)
```



## META



```{r}
queryMETA(vcf)
```


## GT data


Determine what data we have by querrying the FORMAT column.


```{r}
unique(vcf@gt[,1])
```


Subset to only heterozygotes.


```{r}
#vcf@gt[1,1:2]
gt <- extract.gt(vcf)
hets <- is_het(gt)
is.na( vcf@gt[,-1][ !hets ] ) <- TRUE
```



```{r}
vcf@gt[1,1:2]
gt <- extract.gt(vcf)
ad <- extract.gt(vcf, element = 'AD', as.numeric = FALSE)
ad1 <- AD_frequency(ad)
dp <- extract.gt(vcf, element = 'DP', as.numeric = TRUE)
gq <- extract.gt(vcf, element = 'GQ', as.numeric = TRUE)

pl <- extract.gt(vcf, element = 'PL', as.numeric = FALSE)
pl1 <- AD_frequency(pl)


#unlist(lapply(strsplit(ad[,1], split=","), function(x){sort(as.numeric(x), decreasing = TRUE)[1]}))
```



```{r}
par(mar=c(8,4,4,2))
boxplot(dp, las=3, col="#F4A460", ylab="DP")
par(mar=c(5,4,4,2))
```



```{r}
range(pl1[,1], na.rm = TRUE)
seq(min(pl1[,1], na.rm = TRUE), max(pl1[,1], na.rm = TRUE), by=0.02)

hist(pl1[,'P17777us22'], 
#     breaks = seq(min(pl1[,'P17777us22'], na.rm = TRUE), max(pl1[,'P17777us22'], na.rm = TRUE), by=0.01),
     breaks = seq(0.5, 1.0, by=0.01),
     col="#F4A460",
     main = "PL")
```





```{r}

hist(ad1[,'P17777us22'], xlim=c(0,1), breaks=seq(0,1,by=0.02), col="#F4A460", xaxt="n")
axis(side=1, at=c(0, 0.25, 0.333, 0.5, 0.666, 0.75, 1), labels=c(0,"1/4", "1/3", "1/2", "2/3", "3/4", 1), las=3)

mean(ad1[,'P17777us22'], na.rm = TRUE)
median(ad1[,'P17777us22'], na.rm = TRUE)
quantile(ad1[,'P17777us22'], na.rm=TRUE)


library(moments)

skewness(ad1[,'P17777us22'], na.rm = TRUE)

kurtosis(ad1[,'P17777us22'], na.rm = TRUE)

```



```{r}

gtm <- cbind(ad1[,'P17777us22'], dp[,'P17777us22'], gq[,'P17777us22'], pl1[,'P17777us22'])
colnames(gtm) <- c('ad1','dp','gq','pl1')
head(gtm)

apply(gtm, MARGIN=2, class)

gtd <- dist(scale(gtm))

range(gtd, na.rm = TRUE)
gtd[ is.na(gtd) ] <- 0

pco <- cmdscale(gtd)

plot(pco, pch=20, col="#F4A460")

```



```{r}
library(vegan)


gtm2 <- gtm[!apply(gtm, MARGIN=1, function(x){ sum(is.na(x)) == length(x) }),]
INFO2 <- INFO[!apply(gtm, MARGIN=1, function(x){ sum(is.na(x)) == length(x) }),]

INFO2 <- apply(INFO2, MARGIN=2, as.numeric)
INFO2 <- as.data.frame(INFO2)
class(INFO2)
apply(INFO2, MARGIN=2, class)



#gtd <- vegdist(gtm2, method="bray", na.rm = TRUE)


mds1 <- metaMDS(gtm2)

#plot(mds1)

ord <- mds1

plot(ord, type = "n")
points(ord, display = "sites", cex = 0.8, pch=21, col="red", bg="yellow")
text(ord, display = "spec", cex=1.2, col="blue")

#INFO2 <- cbind(NA, NA, INFO2)
colnames(INFO2)

#INFO2[is.na(INFO2)] <- 0
#apply(INFO2, MARGIN=2, function(x){ sum(is.na(x)) })

colnames(INFO2)
INFO2[1:3,1:7]
ord.fit <- envfit(ord, env=INFO2[,-c(7,9,14)], perm=999, na.rm = TRUE)
#ord.fit <- envfit(ord, env=INFO2[,1:6], perm=999, na.rm = TRUE)
#ord.fit <- envfit(ord ~ AC + AF + AN, data=INFO2, perm=999)

plot(ord, type = "n")
points(ord, display = "sites", cex = 0.8, pch=21, col="red", bg="yellow")
text(ord, display = "spec", cex=1.2, col="blue")
#plot(ord.fit, choices = c(1,2))

#nlabs <- labels(ord.fit)
#nlabs$factors <- rep("", times=length(nlabs$factors))

scores(ord.fit, "vectors")
plot(ord.fit, choices = c(1,2), at = c(0,0),
#     labels = nlabs,
     axis = FALSE, p.max = 0.05, col = "#008000", add = TRUE)


```





```{r sample.ord}

library(vcfR)
data(vcfR_test)
x <- vcfR_test


sample.ord <- function(x, sample, print = TRUE, verbose = TRUE){
  require(vegan)
  
  x <- x[,c(1,sample)]
  
  # INFO data
  myINFO <- INFO2df(x)
  myMETA <- metaINFO2df(x)
  
  # Get FORMAT fields
  myFORMAT <- unlist( lapply( queryMETA(x, element="FORMAT="), function(x){x[1]} ) )
  myFORMAT <- unlist( lapply( strsplit(myFORMAT, split="="), function(x){ x[length(x)] } ) )
  myFORMAT <- myFORMAT[ myFORMAT != "GT" ]
  
  
  # Ordination
#  suppressMessages(mds1 <- vegan::metaMDS(gtm2, distance = "bray", k = 2))
#  suppressWarnings(mds1 <- vegan::metaMDS(gtm2, distance = "bray", k = 2))

  # Ordination
  mds1 <- vegan::metaMDS(gtm2, distance = "bray", k = 2)

  # http://stackoverflow.com/a/2501913
#  sink("/dev/null")    # now suppresses
#  mds1 <- vegan::metaMDS(gtm2, distance = "bray", k = 2)
#  sink()

  # Covariates
  ord.fit <- envfit(mds1, env=INFO2[,-c(7,9,14)], perm=999, na.rm = TRUE)
  
  # Plot
  plot(mds1, type = "n")
#  points(mds1, display = "sites", cex = 0.8, pch=21, col="red", bg="yellow")
  points(mds1, display = "sites", cex = 0.8, pch=20, col=rgb(139,69,19, alpha=50, maxColorValue = 255))
  text(mds1, display = "spec", cex=1.2, col="blue")
  
  plot(ord.fit, choices = c(1,2), at = c(0,0),
       axis = FALSE,
#       p.max = 0.05,
       col = "#008000", add = TRUE)
  
  
  return( list( metaMDS = mds1, envfit = ord.fit ) )
}






```





```{r, eval=FALSE}

#png("P17777us22_varord.png", width=800, height=800)

plot(ord, type = "n")
points(ord, display = "sites", cex = 0.8, pch=21, col="red", bg="yellow")
text(ord, display = "spec", cex=1.2, col="blue")

scores(ord.fit, "vectors")
plot(ord.fit, choices = c(1,2), at = c(0,0),
     axis = FALSE, p.max = 0.05, col = "#008000", add = TRUE)

title(main="P17777us22")

#dev.off()


```


